ducktape/src/main/scala/io/github/arainko/ducktape/internal/macros/FallibleCoproductTransformations.scala

```diff
@@ -0,0 +1,84 @@
+package io.github.arainko.ducktape.internal.macros
+
+import io.github.arainko.ducktape.fallible.{ FallibleTransformer, Mode }
+import io.github.arainko.ducktape.function.FunctionArguments
+import io.github.arainko.ducktape.internal.modules.MaterializedConfiguration.*
+import io.github.arainko.ducktape.internal.modules.*
+
+import scala.deriving.*
+import scala.quoted.*
+
+private[ducktape] object FallibleCoproductTransformations {
+
+  def transform[F[+x]: Type, Source: Type, Dest: Type](
+    Source: Expr[Mirror.SumOf[Source]],
+    Dest: Expr[Mirror.SumOf[Dest]],
+    F: Expr[Mode[F]],
+    sourceValue: Expr[Source]
+  )(using Quotes): Expr[F[Dest]] = {
+    given Cases.Source = Cases.Source.fromMirror(Source)
+    given Cases.Dest = Cases.Dest.fromMirror(Dest)
+
+    val ifBranches = coproductBranches[F, Source, Dest](sourceValue, Cases.source.value, F)
+    ifStatement(ifBranches).asExprOf[F[Dest]]
+  }
+
+  private def coproductBranches[F[+x]: Type, Source: Type, Dest: Type](
+    sourceValue: Expr[Source],
+    sourceCases: List[Case],
+    F: Expr[Mode[F]]
+  )(using Quotes, Cases.Dest) = {
+    import quotes.reflect.*
+
+    sourceCases.map { source =>
+      source -> Cases.dest
+        .get(source.name)
+        .getOrElse(Failure.emit(Failure.NoChildMapping(source.name, summon[Type[Dest]])))
+    }.map { (source, dest) =>
+      val cond = IsInstanceOf(sourceValue, source.tpe)
+
+      (source.tpe -> dest.tpe) match {
+        case '[src] -> '[dest] =>
+          val value =
+            source.fallibleTransformerTo[F](dest).map {
+              case '{ $transformer: FallibleTransformer[F, src, dest] } =>
+                '{
+                  val castedSource = $sourceValue.asInstanceOf[src]
+                  val t = $transformer
+                  t.transform(castedSource)
+                }.asExprOf[F[Dest]]
+            } match {
+              case Right(value) => value
+              case Left(explanation) =>
+                dest.materializeSingleton.map { singleton =>
+                  '{ $F.pure[dest]($singleton.asInstanceOf[dest]) }
+                }
+                  .getOrElse(
+                    Failure.emit(
+                      Failure.CannotTransformCoproductCaseFallible(summon[Type[F]], source.tpe, dest.tpe, explanation)
+                    )
+                  )
+            }
+          IfBranch(cond, value)
+      }
+    }
+  }
+
+  private def ifStatement(using Quotes)(branches: List[IfBranch]): quotes.reflect.Term = {
```

Comment:
So we use duplicates of this, `IfBranch` and `IsInstanceOf` inside `CoproductTransformations` as well, would you mind extracting these into a separate and reusable object? eg. `IfStatement` that'd house `IfBranch` and other things related to this inside

Reply:
I've extracted the `IfBranch` logic into its own reusable object `ExhaustiveCoproductMatching`. Maybe I've went a bit to far there. What do you say?